### ISSUE TITLE
[PATCH] Add tests for stderr-only in IPC::Cmd

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9051]
Migrated from [rt.perl.org#46111](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=46111) (status was 'resolved')

Searchable as RT46111$

**Comment by p5pRT** [comment]
### From @doughera88

The following patch adds tests for commands that print only to stderr to 
the IPC&#8203;::Cmd test suite\.

These tests currently fail for me under Solaris 8 with Sun's cc compiler\. 
\(And this failure ultimately shows up as a failure in Archive&#8203;::Extract 
when it tries to parse the stderr output of bzip2 \-\-version\)\.


<details><summary>Inline Patch</summary>

```diff
diff -ru perl-current/lib/IPC/Cmd/t/01_IPC-Cmd.t perl-andy/lib/IPC/Cmd/t/01_IPC-Cmd.t
--- perl-current/lib/IPC/Cmd/t/01_IPC-Cmd.t	Tue Nov 28 05:45:22 2006
+++ perl-andy/lib/IPC/Cmd/t/01_IPC-Cmd.t	Fri Oct  5 13:26:32 2007
@@ -35,7 +35,7 @@
     ok( !can_run('10283lkjfdalskfjaf'), q[Not found non-existant binary] );
 }
 
-### run tests
+### run tests that print only to stdout
 {   ### list of commands and regexes matching output ###
     my $map = [
         # command                                    # output regex
@@ -45,6 +45,7 @@
         [ [$^X,qw[-eprint+42 |], $^X, qw|-neprint|], qr/42/,            ],
     ];
 
+    diag( "Running tests that print only to stdout" ) if $Verbose;
     ### for each configuarion
     for my $pref ( @Prefs ) {
         diag( "Running config: IPC::Run: $pref->[0] IPC::Open3: $pref->[1]" )
@@ -108,6 +109,79 @@
     }
 }
 
+### run tests that print only to stderr
+{   ### list of commands and regexes matching output ###
+    my $map = [
+        # command                                    # output regex
+        [ "$^X -ewarn+42",                          qr/^42 /, ],
+        [ [$^X, '-ewarn+42'],                       qr/^42 /, ],
+    ];
+
+    diag( "Running tests that print only to stderr" ) if $Verbose;
+    ### for each configuarion
+    for my $pref ( @Prefs ) {
+        diag( "Running config: IPC::Run: $pref->[0] IPC::Open3: $pref->[1]" )
+            if $Verbose;
+
+        $IPC::Cmd::USE_IPC_RUN    = $IPC::Cmd::USE_IPC_RUN      = $pref->[0];
+        $IPC::Cmd::USE_IPC_OPEN3  = $IPC::Cmd::USE_IPC_OPEN3    = $pref->[1];
+
+        ### for each command
+        for my $aref ( @$map ) {
+            my $cmd                 = $aref->[0];
+            my $regex               = $aref->[1];
+
+            my $pp_cmd = ref $cmd ? "@$cmd" : "$cmd";
+            diag( "Running '$pp_cmd' as " . (ref $cmd ? "ARRAY" : "SCALAR") )
+                if $Verbose;
+
+            ### in scalar mode
+            {   diag( "Running stderr command in scalar mode" ) if $Verbose;
+                my $buffer;
+                my $ok = run( command => $cmd, buffer => \$buffer );
+
+                ok( $ok,        "Ran stderr command succesfully in scalar mode." );
+
+                SKIP: {
+		    # No buffers are expected if neither IPC::Run nor IPC::Open3 is used.
+                    skip "No buffers available", 1
+                                unless $Class->can_capture_buffer;
+
+                    like( $buffer, $regex,
+                                "   Buffer filled properly from stderr" );
+                }
+            }
+
+            ### in list mode
+            {   diag( "Running stderr command in list mode" ) if $Verbose;
+                my @list = run( command => $cmd );
+                ok( $list[0],   "Ran stderr command successfully in list mode." );
+                ok( !$list[1],  "   No error code set" );
+
+                my $list_length = $Class->can_capture_buffer ? 5 : 2;
+                is( scalar(@list), $list_length,
+                                "   Output list has $list_length entries" );
+
+                SKIP: {
+		    # No buffers are expected if neither IPC::Run nor IPC::Open3 is used.
+                    skip "No buffers available", 6
+                                unless $Class->can_capture_buffer;
+
+                    ### the last 3 entries from the RV, are they array refs?
+                    isa_ok( $list[$_], 'ARRAY' ) for 2..4;
+
+                    like( "@{$list[2]}", $regex,
+                                "   Combined buffer holds output" );
+
+                    is( scalar( @{$list[3]} ), 0,
+                                    "   Stdout buffer empty" );
+                    like( "@{$list[4]}", qr/$regex/,
+                            "   Stderr buffer filled" );
+                }
+            }
+        }
+    }
+}
 
 ### test failures
 {   ### for each configuarion

-- 
```
</details>

&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu

**Comment by p5pRT** [comment]
### From @doughera88

The problems exposed by this patch should be solved by \[perl \#46293\]\.  
This patch should still be applied \-\- it just adds tests\.

\-\- 
&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu Oct 11 07&#8203;:15&#8203;:09 2007\, doughera wrote&#8203;:
> The problems exposed by this patch should be solved by \[perl \#46293\]\.  
> This patch should still be applied \-\- it just adds tests\.
> 

Per instructions in 'perlhack'\, I opened a ticket for these tests in
IPC\-Cmd's rt\.cpan\.org queue&#8203;:

https://rt.cpan.org/Public/Bug/Display.html?id=73270

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Dec 14 18&#8203;:26&#8203;:02 2011\, jkeenan wrote&#8203;:
> On Thu Oct 11 07&#8203;:15&#8203;:09 2007\, doughera wrote&#8203;:
> > The problems exposed by this patch should be solved by \[perl \#46293\]\.  
> > This patch should still be applied \-\- it just adds tests\.
> > 
> 
> Per instructions in 'perlhack'\, I opened a ticket for these tests in
> IPC\-Cmd's rt\.cpan\.org queue&#8203;:
> 
> https://rt.cpan.org/Public/Bug/Display.html?id=73270
> 
> Thank you very much\.
> Jim Keenan

https://rt.cpan.org/Ticket/Display.html?id=73270 has been resolved\, so
this ticket can also be marked resolved\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
